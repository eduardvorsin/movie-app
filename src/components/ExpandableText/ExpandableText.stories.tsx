import type { Meta, StoryObj } from '@storybook/react';
import ExpandableText from './ExpandableText';

const meta: Meta<typeof ExpandableText> = {
	title: 'components/ExpandableText',
	tags: ['autodocs'],
	argTypes: {
		className: {
			description: 'Additional classes for ExpandableText'
		},
		visibleRowsCount: {
			description: 'The number of visible rows in the non-expanded state. It can take values from 1 to 6',
		},
		children: {
			description: 'ExpandableText contents',
		},
		defaultExpanded: {
			description: 'Allows you to set whether the text will be open or not when it first appears. By default, part of the text is hidden.',
		},
		testId: {
			description: 'Id for testing the component',
		}
	},
	component: ExpandableText,
	parameters: {
		docs: {
			description: {
				component: 'This component is used to provide the user with the ability to expand or collapse additional text in the interface of a website or mobile application. It can be useful when you need to save space on a page, hide details, improve readability on mobile devices.',
			},
		},
	},
}

export default meta;
type Story = StoryObj<typeof ExpandableText>;

export const Default: Story = {
	args: {
		className: 'max-w-[15rem]',
		children: 'Lorem ipsum dolor sit amet consectetur adipisicing elit. Commodi, dolore voluptate delectus facilis libero odio. Nisi doloremque nulla dolore eius adipisci? Reiciendis aliquam hic est laboriosam inventore expedita mollitia quidem?'
	},
	parameters: {
		docs: {
			description: {
				story: 'Default state of the component',
			},
		},
	},
};

export const DefaultExpanded: Story = {
	args: {
		className: 'max-w-[15rem]',
		children: ['Lorem ipsum dolor sit amet consectetur adipisicing elit. Commodi, dolore voluptate delectus facilis libero odio.', ' Nisi doloremque nulla dolore eius adipisci?', ' Reiciendis aliquam hic est laboriosam inventore expedita mollitia quidem?'],
		defaultExpanded: true,
	},
	parameters: {
		docs: {
			description: {
				story: 'Allows you to set whether the text will be in the expanded or collapsed state when it first appears.',
			},
		},
	},
};