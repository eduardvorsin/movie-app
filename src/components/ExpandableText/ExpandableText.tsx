'use client';
import { MouseEventHandler, useLayoutEffect, useRef, useState } from 'react';
import Button from '@/components/Button/Button';
import { useTranslation } from '@/i18n/client';
import { useParams } from 'next/navigation';
import { Locales, fallbackLng } from '@/i18n/settings';
import { GeneralProps } from '@/types/shared';

type Props = {
	defaultExpanded?: boolean,
	children: string | string[],
	visibleRowsCount?: 1 | 2 | 3 | 4 | 5 | 6,
} & GeneralProps;

const lineClamps = {
	1: 'line-clamp-1',
	2: 'line-clamp-2',
	3: 'line-clamp-3',
	4: 'line-clamp-4',
	5: 'line-clamp-5',
	6: 'line-clamp-6',
} as const;

export default function ExpandableText({
	className,
	children,
	testId,
	defaultExpanded,
	visibleRowsCount = 2,
	...props
}: Props) {
	const lang = useParams()?.lang as Locales ?? fallbackLng;
	const { t } = useTranslation(lang);
	const [isExpanded, setIsExpanded] = useState<boolean>(defaultExpanded ?? false);

	const textRef = useRef<HTMLParagraphElement | null>(null);
	const [shouldShowButton, setShouldShowButton] = useState<boolean>(false);

	const clickHandler: MouseEventHandler<HTMLButtonElement> = () => {
		setIsExpanded((state) => !state);
	};

	useLayoutEffect(() => {
		if (textRef.current) {
			const height = textRef.current.scrollHeight;
			const lineHeight = parseInt(window.getComputedStyle(textRef.current).lineHeight);
			const numberOfRows = height / lineHeight;

			setShouldShowButton(numberOfRows > visibleRowsCount);
		}
	}, [visibleRowsCount]);

	const textClasses = [
		'text-neutral-1000 dark:text-dark-neutral-1000',
		!isExpanded ? lineClamps[visibleRowsCount] : '',
	].join(' ');

	return (
		<div
			className={className}
			data-testid={testId}
			{...props}
		>
			{typeof children === 'string' ?
				(
					<p
						ref={textRef}
						className={textClasses}
					>
						{children}
					</p>
				)
				:
				(
					<div
						ref={textRef}
						className={textClasses}
					>
						{children.map((paragraphText, index) => (
							<p key={index} className='mb-2 last:mb-0'>{paragraphText}</p>
						))}
					</div>
				)
			}
			{shouldShowButton && (
				<Button
					appearance='ghost'
					className='mt-1 py-2 [&]:px-0'
					onClick={clickHandler}
				>
					{isExpanded ?
						t('expandableText.collapseButton') :
						t('expandableText.expandButton')
					}
				</Button>
			)}
		</div>
	);
}