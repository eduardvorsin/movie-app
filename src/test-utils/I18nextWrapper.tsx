import React, { ReactNode } from 'react';
import i18next from '@/i18n/client';
import { I18nextProvider } from 'react-i18next';

type Props = {
	children: ReactNode,
}

const I18nextWrapper = ({
	children,
}: Props) => {
	return (
		<I18nextProvider i18n={i18next}>
			{children}
		</I18nextProvider>
	);
};

export default I18nextWrapper;