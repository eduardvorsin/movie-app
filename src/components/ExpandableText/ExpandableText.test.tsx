import { render, screen } from '@testing-library/react';
import ExpandableText from './ExpandableText';
import i18next from '@/i18n/client';
import I18nextWrapper from '@/test-utils/I18nextWrapper';

jest.mock<typeof import('next/navigation')>('next/navigation');

describe('ExpandableText tests', () => {
	beforeEach(() => {
		Object.defineProperty(HTMLElement.prototype, 'scrollHeight', {
			configurable: true,
			value: 200,
		});
		jest.spyOn(window, 'getComputedStyle').mockReturnValue(
			{ lineHeight: '20px' } as CSSStyleDeclaration
		);
	});

	it('is rendered correctly', async () => {
		const dictionary = {
			collapseButton: 'Collapse',
			expandButton: 'Expand',
		};

		render(
			<ExpandableText
				dictionary={dictionary}
				testId='test-expandable-text'
			>
				Lorem ipsum dolor sit amet consectetur adipisicing elit. Quo repellat adipisci et, eligendi ducimus accusamus consectetur, labore iusto esse, ratione pariatur laboriosam. Repudiandae, provident aspernatur. Adipisci nisi veniam incidunt tempore
			</ExpandableText>,
			{ wrapper: I18nextWrapper }
		);

		await screen.findByTestId<HTMLDivElement>('test-expandable-text');
		expect(screen.getByTestId<HTMLDivElement>('test-expandable-text')).toBeInTheDocument();
	});

	it('is a basic snapshot', async () => {
		const dictionary = {
			collapseButton: 'Collapse',
			expandButton: 'Expand',
		};

		render(
			<ExpandableText
				dictionary={dictionary}
				testId='test-expandable-text'
			>
				Lorem ipsum dolor sit amet consectetur adipisicing elit. Quo repellat adipisci et, eligendi ducimus accusamus consectetur, labore iusto esse, ratione pariatur laboriosam. Repudiandae, provident aspernatur. Adipisci nisi veniam incidunt tempore
			</ExpandableText>
		);

		await screen.findByTestId<HTMLDivElement>('test-expandable-text');
		expect(screen.getByTestId<HTMLDivElement>('test-expandable-text')).toMatchSnapshot();
	});

	it('is a snapshot with defaultExpanded set to true', async () => {
		const dictionary = {
			collapseButton: 'Collapse',
			expandButton: 'Expand',
		};

		render(
			<ExpandableText
				dictionary={dictionary}
				testId='test-expandable-text'
				defaultExpanded
			>
				Lorem ipsum dolor sit amet consectetur adipisicing elit. Quo repellat adipisci et, eligendi ducimus accusamus consectetur, labore iusto esse, ratione pariatur laboriosam. Repudiandae, provident aspernatur. Adipisci nisi veniam incidunt tempore
			</ExpandableText>
		);

		await screen.findByTestId<HTMLDivElement>('test-expandable-text');
		expect(screen.getByTestId<HTMLDivElement>('test-expandable-text')).toMatchSnapshot();
	});
});

describe('ExpandableText integration tests', () => {
	beforeEach(() => {
		Object.defineProperty(HTMLElement.prototype, 'scrollHeight', {
			configurable: true,
			value: 200,
		});
		jest.spyOn(window, 'getComputedStyle').mockReturnValue(
			{ lineHeight: '20px' } as CSSStyleDeclaration
		);
	});

	it('localization into English works correctly', async () => {
		await i18next.changeLanguage('en');
		const translation = new RegExp(i18next.t('expandableText.expandButton'));

		const dictionary = {
			collapseButton: i18next.t('expandableText.collapseButton'),
			expandButton: i18next.t('expandableText.expandButton'),
		};

		render(
			<ExpandableText
				dictionary={dictionary}
				testId='test-expandable-text'
			>
				Lorem ipsum dolor sit amet consectetur adipisicing elit. Quo repellat adipisci et, eligendi ducimus accusamus consectetur, labore iusto esse, ratione pariatur laboriosam. Repudiandae, provident aspernatur. Adipisci nisi veniam incidunt tempore
			</ExpandableText>,
			{ wrapper: I18nextWrapper }
		);

		await screen.findByTestId<HTMLDivElement>('test-expandable-text');
		expect(screen.getByText<HTMLSpanElement>(translation)).toBeInTheDocument();
	});

	it('localization into Russian works correctly', async () => {
		await i18next.changeLanguage('ru');
		const translation = new RegExp(i18next.t('expandableText.expandButton'));

		const dictionary = {
			collapseButton: i18next.t('expandableText.collapseButton'),
			expandButton: i18next.t('expandableText.expandButton'),
		};

		render(
			<ExpandableText
				dictionary={dictionary}
				testId='test-expandable-text'
			>
				Lorem ipsum dolor sit amet consectetur adipisicing elit. Quo repellat adipisci et, eligendi ducimus accusamus consectetur, labore iusto esse, ratione pariatur laboriosam. Repudiandae, provident aspernatur. Adipisci nisi veniam incidunt tempore
			</ExpandableText>,
			{ wrapper: I18nextWrapper }
		);

		await screen.findByTestId<HTMLDivElement>('test-expandable-text');
		expect(screen.getByText<HTMLSpanElement>(translation)).toBeInTheDocument();
	});

	it('is a snapshot with English localization', async () => {
		await i18next.changeLanguage('en');

		const dictionary = {
			collapseButton: i18next.t('expandableText.collapseButton'),
			expandButton: i18next.t('expandableText.expandButton'),
		};

		render(
			<ExpandableText
				dictionary={dictionary}
				testId='test-expandable-text'
			>
				Lorem ipsum dolor sit amet consectetur adipisicing elit. Quo repellat adipisci et, eligendi ducimus accusamus consectetur, labore iusto esse, ratione pariatur laboriosam. Repudiandae, provident aspernatur. Adipisci nisi veniam incidunt tempore
			</ExpandableText>,
			{ wrapper: I18nextWrapper }
		);

		await screen.findByTestId<HTMLDivElement>('test-expandable-text');
		expect(screen.getByTestId<HTMLDivElement>('test-expandable-text')).toMatchSnapshot();
	});

	it('is a snapshot with Russian localization', async () => {
		await i18next.changeLanguage('ru');

		const dictionary = {
			collapseButton: i18next.t('expandableText.collapseButton'),
			expandButton: i18next.t('expandableText.expandButton'),
		};

		render(
			<ExpandableText
				dictionary={dictionary}
				testId='test-expandable-text'
			>
				Lorem ipsum dolor sit amet consectetur adipisicing elit. Quo repellat adipisci et, eligendi ducimus accusamus consectetur, labore iusto esse, ratione pariatur laboriosam. Repudiandae, provident aspernatur. Adipisci nisi veniam incidunt tempore
			</ExpandableText>,
			{ wrapper: I18nextWrapper }
		);

		await screen.findByTestId<HTMLDivElement>('test-expandable-text');
		expect(screen.getByTestId<HTMLDivElement>('test-expandable-text')).toMatchSnapshot();
	});
});