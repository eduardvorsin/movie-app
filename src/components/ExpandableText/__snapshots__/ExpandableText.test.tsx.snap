// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpandableText integration tests is a snapshot with English localization 1`] = `
<div
  data-testid="test-expandable-text"
>
  <p
    class="text-neutral-1000 dark:text-dark-neutral-1000 transition-colors duration-150 line-clamp-2"
  >
    Lorem ipsum dolor sit amet consectetur adipisicing elit. Quo repellat adipisci et, eligendi ducimus accusamus consectetur, labore iusto esse, ratione pariatur laboriosam. Repudiandae, provident aspernatur. Adipisci nisi veniam incidunt tempore
  </p>
  <button
    class="items-center font-medium rounded-1 text-center cursor-pointer min-w[2.25rem] leading-2 inline-flex transition-colors duration-150 text-100  text-dark-neutral-0 enabled:hover:text-dark-neutral-400 enabled:active:text-dark-neutral-600 dark:text-neutral-200 dark:enabled:hover:text-neutral-400 dark:enabled:active:text-neutral-600 py-2 px-4 relative  mt-1 py-2 [&]:px-0"
  >
    <span
      class="inline-flex items-center"
    >
      Expand
    </span>
  </button>
</div>
`;

exports[`ExpandableText integration tests is a snapshot with Russian localization 1`] = `
<div
  data-testid="test-expandable-text"
>
  <p
    class="text-neutral-1000 dark:text-dark-neutral-1000 transition-colors duration-150 line-clamp-2"
  >
    Lorem ipsum dolor sit amet consectetur adipisicing elit. Quo repellat adipisci et, eligendi ducimus accusamus consectetur, labore iusto esse, ratione pariatur laboriosam. Repudiandae, provident aspernatur. Adipisci nisi veniam incidunt tempore
  </p>
  <button
    class="items-center font-medium rounded-1 text-center cursor-pointer min-w[2.25rem] leading-2 inline-flex transition-colors duration-150 text-100  text-dark-neutral-0 enabled:hover:text-dark-neutral-400 enabled:active:text-dark-neutral-600 dark:text-neutral-200 dark:enabled:hover:text-neutral-400 dark:enabled:active:text-neutral-600 py-2 px-4 relative  mt-1 py-2 [&]:px-0"
  >
    <span
      class="inline-flex items-center"
    >
      Развернуть
    </span>
  </button>
</div>
`;

exports[`ExpandableText tests is a basic snapshot 1`] = `
<div
  data-testid="test-expandable-text"
>
  <p
    class="text-neutral-1000 dark:text-dark-neutral-1000 transition-colors duration-150 line-clamp-2"
  >
    Lorem ipsum dolor sit amet consectetur adipisicing elit. Quo repellat adipisci et, eligendi ducimus accusamus consectetur, labore iusto esse, ratione pariatur laboriosam. Repudiandae, provident aspernatur. Adipisci nisi veniam incidunt tempore
  </p>
  <button
    class="items-center font-medium rounded-1 text-center cursor-pointer min-w[2.25rem] leading-2 inline-flex transition-colors duration-150 text-100  text-dark-neutral-0 enabled:hover:text-dark-neutral-400 enabled:active:text-dark-neutral-600 dark:text-neutral-200 dark:enabled:hover:text-neutral-400 dark:enabled:active:text-neutral-600 py-2 px-4 relative  mt-1 py-2 [&]:px-0"
  >
    <span
      class="inline-flex items-center"
    >
      Expand
    </span>
  </button>
</div>
`;

exports[`ExpandableText tests is a snapshot with defaultExpanded set to true 1`] = `
<div
  data-testid="test-expandable-text"
>
  <p
    class="text-neutral-1000 dark:text-dark-neutral-1000 transition-colors duration-150 "
  >
    Lorem ipsum dolor sit amet consectetur adipisicing elit. Quo repellat adipisci et, eligendi ducimus accusamus consectetur, labore iusto esse, ratione pariatur laboriosam. Repudiandae, provident aspernatur. Adipisci nisi veniam incidunt tempore
  </p>
  <button
    class="items-center font-medium rounded-1 text-center cursor-pointer min-w[2.25rem] leading-2 inline-flex transition-colors duration-150 text-100  text-dark-neutral-0 enabled:hover:text-dark-neutral-400 enabled:active:text-dark-neutral-600 dark:text-neutral-200 dark:enabled:hover:text-neutral-400 dark:enabled:active:text-neutral-600 py-2 px-4 relative  mt-1 py-2 [&]:px-0"
  >
    <span
      class="inline-flex items-center"
    >
      Collapse
    </span>
  </button>
</div>
`;
